src/v/cloud_topics/level_one/compaction/filter.cc

```diff
@@ -22,17 +24,46 @@
 namespace cloud_topics::l1 {
 
 compaction_filter::compaction_filter(
-  compaction::sliding_window_reducer::sink& sink,
+  compaction_sink& sink,
   const compaction::key_offset_map& map,
-  model::ntp ntp)
-  : filter(sink, std::move(ntp))
-  , _map(map) {}
+  model::ntp ntp,
+  const offset_interval_set& removable_tombstone_ranges,
+  kafka::offset start_offset,
+  kafka::offset last_offset)
+  : filter(
+      sink,
+      std::move(ntp),
+      kafka::offset_cast(start_offset),
+      kafka::offset_cast(last_offset))
+  , _ct_sink(sink)
+  , _map(map)
+  , _removable_tombstone_ranges(removable_tombstone_ranges) {}
+
+ss::future<bool> compaction_filter::should_keep(
+  const model::record_batch& b, const model::record& r) const {
+    if (r.is_tombstone()) {
+        auto o = model::offset_cast(
+          b.base_offset() + model::offset_delta(r.offset_delta()));
+        if (_removable_tombstone_ranges.contains(o)) {
+            ++_stats.expired_tombstones_discarded;
+            co_return false;
+        }
+    }
+
+    auto keep = co_await compaction::is_latest_record_for_key(_map, b, r);
+
+    if (r.is_tombstone() && keep) {
+        _ct_sink.set_range_has_tombstones();
```

Comment:
It feels a bit unnatural to be mutating the sink from this const method. I know it's a bit of plumbing, but WDYT about, as a part of `filter_batch_with_offset_deltas()` this returned whether the resulting batch contains tombstones so we can do this setting in the sink?

I think the logic translates naturally to other compaction implementations, not just cloud topics, so it does feel like baking it into the interfaces seems like a step in the right direction